zebra-state/src/service/read/difficulty.rs

```diff
@@ -102,92 +98,102 @@ where
         .checked_add_signed(Duration::seconds(BLOCK_MAX_TIME_SINCE_MEDIAN))
         .expect("median time plus a small constant is far below i64::MAX");
 
-    let current_system_time = current_system_time
+    let cur_time = cur_time
         .timestamp()
         .clamp(min_time.timestamp(), max_time.timestamp());
 
-    let mut current_system_time = Utc.timestamp_opt(current_system_time, 0).single().expect(
+    let cur_time = Utc.timestamp_opt(cur_time, 0).single().expect(
         "clamping a timestamp between two valid times can't make it invalid, and \
              UTC never has ambiguous time zone conversions",
     );
 
     // Now that we have a valid time, get the difficulty for that time.
-    let mut difficulty_adjustment = AdjustedDifficulty::new_from_header_time(
-        current_system_time,
+    let difficulty_adjustment = AdjustedDifficulty::new_from_header_time(
+        cur_time,
         tip.0,
         network,
         relevant_data.iter().cloned(),
     );
 
+    let mut result = GetBlockTemplateChainInfo {
+        tip,
+        expected_difficulty: difficulty_adjustment.expected_difficulty_threshold(),
+        min_time,
+        cur_time,
+        max_time,
+    };
+
+    adjust_difficulty_and_time_for_testnet(&mut result, network, tip.0, relevant_data);
+
+    result
+}
+
+/// Adjust the difficulty and time for the testnet minimum difficulty rule.
+fn adjust_difficulty_and_time_for_testnet(
+    result: &mut GetBlockTemplateChainInfo,
+    network: Network,
+    previous_block_height: Height,
+    relevant_data: Vec<(CompactDifficulty, DateTime<Utc>)>,
+) {
+    if network == Network::Mainnet {
+        return;
+    }
+
     // On testnet, changing the block time can also change the difficulty,
     // due to the minimum difficulty consensus rule:
     // > if the block time of a block at height height ≥ 299188
     // > is greater than 6 * PoWTargetSpacing(height) seconds after that of the preceding block,
     // > then the block is a minimum-difficulty block.
     //
-    // In this case, we adjust the min_time and cur_time to the first minimum difficulty time.
+    // The max time is always a minimum difficulty block, because the minimum difficulty
+    // gap is 15 minutes, but the maximum gap is 90 minutes. This means that testnet blocks
+    // have two valid time ranges with different difficulties:
+    // * 1s - 15m: standard difficulty
```

Comment:
> is greater than 6 * PoWTargetSpacing(height) seconds after that of the preceding block

Would the standard difficulty time range then be 1s - 9m?

Reply:
It's actually 7.5 minutes, the 15 is from before the Blossom target spacing change:
https://github.com/ZcashFoundation/zebra/blob/5c4d3e1f5ca88032b4d81f8839997a642475aabd/zebra-chain/src/parameters/network_upgrade.rs#L200

I'll go fix that now.

Reply:
Fixed in e26c1c4